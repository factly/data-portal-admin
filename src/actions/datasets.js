import axios from '../utils/axios';
import {
  ADD_DATASET,
  ADD_DATASET_FORMAT,
  REMOVE_DATASET_FORMAT,
  ADD_DATASETS,
  SET_DATASET_LOADING,
  SET_DATASET_REQUEST,
  SET_DATASET_IDS,
  RESET_DATASET,
  DATASET_API,
} from '../constants/datasets';
import { addCurrencies } from './currencies';
import { addTags } from './tags';
import { addMedium } from './media';
import { getIds, buildObjectOfItems, getValues, deleteKeys } from '../utils/objects';

export const loadDatasets = (page = 1, limit = 5) => {
  return async (dispatch, getState) => {
    const {
      datasets: { req },
    } = getState();

    let ids;
    for (let item of req) {
      if (item.page === page && item.limit === limit) {
        ids = [...item.ids];
      }
    }

    if (ids) {
      dispatch(setDatasetIds(ids));
      return;
    }

    dispatch(setLoading(true));

    const response = await axios({
      url: `${DATASET_API}?page=${page}&limit=${limit}`,
      method: 'get',
    });

    const { nodes, total } = response.data;
    const currentPageIds = getIds(nodes);
    const currentReq = { page: page, limit: limit, ids: currentPageIds };
    dispatch(setDatasetRequest(currentReq, total));
    dispatch(addDatasets(nodes));
    dispatch(setDatasetIds(currentPageIds));

    dispatch(setLoading(false));
  };
};

export const createDataset = (data) => {
  return async (dispatch) => {
    dispatch(setLoading(true));

    const response = await axios({
      url: DATASET_API,
      method: 'post',
      data: data,
    });

    dispatch(addDataset(response.data));
    dispatch(setLoading(false));

    return response.data;
  };
};

export const createDatasetFormat = (datasetId, data) => {
  return async (dispatch) => {
    dispatch(setLoading(true));

    const response = await axios({
      url: `${DATASET_API}/${datasetId}/format`,
      method: 'post',
      data: data,
    });

    dispatch(addDatasetFormat(datasetId, response.data));
    dispatch(setLoading(false));
  };
};

export const updateDataset = (id, data) => {
  return async (dispatch) => {
    let url = `${DATASET_API}/${id}`;

    const response = await axios({
      url: url,
      method: 'put',
      data: data,
    });

    dispatch(addDataset(response.data));

    return response.data;
  };
};

export const deleteDataset = (id) => {
  return async (dispatch, getState) => {
    let url = `${DATASET_API}/${id}`;

    dispatch(setLoading(true));

    await axios({
      url: url,
      method: 'delete',
    });

    dispatch(resetDataset());
    dispatch(setLoading(false));
  };
};

export const deleteDatasetFormat = (datasetId, id) => {
  return async (dispatch) => {
    let url = `${DATASET_API}/${datasetId}/format/${id}`;

    dispatch(setLoading(true));

    await axios({
      url: url,
      method: 'delete',
    });

    dispatch(removeDatasetFormat(datasetId, id));
    dispatch(setLoading(false));
  };
};

export const getDataset = (id) => {
  return async (dispatch, getState) => {
    const {
      datasets: { items },
    } = getState();

    if (items[id]) {
      return;
    }

    dispatch(setLoading(true));

    const response = await axios({
      url: `${DATASET_API}/${id}`,
      method: 'get',
    });
    dispatch(addDataset(response.data));

    dispatch(setLoading(false));
  };
};

export const setLoading = (loading) => {
  return {
    type: SET_DATASET_LOADING,
    payload: { loading },
  };
};

export const addDataset = (dataset) => (dispatch) => {
  const currencies = getValues([dataset], 'currency');
  dispatch(addCurrencies(currencies));

  const medium = getValues([dataset], 'featured_medium');
  dispatch(addMedium(medium));

  const tags = getValues([dataset], 'tags');
  dispatch(addTags(tags));

  dataset.tags = getIds(dataset.tags);

  dispatch({
    type: ADD_DATASET,
    payload: { dataset: deleteKeys([dataset], ['currency', 'featured_medium'])[0] },
  });
};

export const addDatasets = (datasets) => (dispatch) => {
  const currencies = getValues(datasets, 'currency');
  dispatch(addCurrencies(currencies));

  const medium = getValues(datasets, 'featured_medium');
  dispatch(addMedium(medium));

  const tags = getValues(datasets, 'tags');
  dispatch(addTags(tags));

  datasets.forEach((dataset) => {
    dataset.tags = getIds(dataset.tags);
  });

  dispatch({
    type: ADD_DATASETS,
    payload: {
      datasets: buildObjectOfItems(deleteKeys(datasets, ['currency', 'featured_medium'])),
    },
  });
};

export const addDatasetFormat = (datasetId, datasetFormat) => {
  return {
    type: ADD_DATASET_FORMAT,
    payload: { datasetId, datasetFormat },
  };
};

export const removeDatasetFormat = (datasetId, datasetFormatId) => {
  return {
    type: REMOVE_DATASET_FORMAT,
    payload: { datasetId, datasetFormatId },
  };
};

export const setDatasetRequest = (req, total) => {
  return {
    type: SET_DATASET_REQUEST,
    payload: { req, total },
  };
};

export const setDatasetIds = (ids) => {
  return {
    type: SET_DATASET_IDS,
    payload: { ids },
  };
};

export const resetDataset = () => {
  return {
    type: RESET_DATASET,
  };
};
