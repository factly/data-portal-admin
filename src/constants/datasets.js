export const LOADING_DATASETS = 'LOADING_DATASETS';
export const LOAD_DATASETS_SUCCESS = 'LOAD_DATASETS_SUCCESS';
export const SET_DATASETS_LIST_TOTAL = 'SET_DATASETS_LIST_TOTAL';
export const ADD_DATASETS_LIST_REQUEST = 'ADD_DATASETS_LIST_REQUEST';
export const SET_DATASETS_LIST_CURRENT_PAGE = 'SET_DATASETS_LIST_CURRENT_PAGE';
export const LOAD_DATASETS_FAILURE = 'LOAD_DATASETS_FAILURE';
export const CREATING_DATASET = 'CREATING_DATASET';
export const CREATE_DATASET_SUCCESS = 'CREATE_DATASET_SUCCESS';
export const CREATE_DATASET_FAILURE = 'CREATE_DATASET_FAILURE';
export const UPDATING_DATASET = 'UPDATING_DATASET';
export const UPDATE_DATASET_SUCCESS = 'UPDATE_DATASET_SUCCESS';
export const UPDATE_DATASET_FAILURE = 'UPDATE_DATASET_FAILURE';
export const DELETING_DATASET = 'DELETING_DATASET';
export const DELETE_DATASET_SUCCESS = 'DELETE_DATASET_SUCCESS';
export const DELETE_DATASET_FAILURE = 'DELETE_DATASET_FAILURE';

export const baseUrl = `${process.env.REACT_APP_API_URL}/datasets`;
