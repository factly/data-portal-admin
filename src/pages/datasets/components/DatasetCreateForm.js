import React, { useState } from 'react';
import { Steps } from 'antd';

import DatasetForm from './DatasetForm';
import DatasetFormatForm from './DatasetFormatForm';

const { Step } = Steps;

const DatasetCreateForm = ({ dataset, datasetFormats, onSubmitDataset, onSubmitDatasetFormat }) => {
  const [currentStep, setCurrentStep] = useState(0);
  const [datasetId, setDatasetId] = useState(dataset?.id);

  const next = () => {
    setCurrentStep(currentStep + 1);
  };
  const prev = () => {
    setCurrentStep(currentStep - 1);
  };

  const steps = [
    {
      title: 'Add Dataset Details',
      content: (
        <DatasetForm
          data={dataset}
          onSubmit={onSubmitDataset}
          setDatasetId={setDatasetId}
          next={next}
        />
      ),
    },
    {
      title: 'Add Datasets',
      content: (
        <DatasetFormatForm
          datasetFormats={datasetFormats}
          datasetId={datasetId}
          onSubmit={(data) => onSubmitDatasetFormat(datasetId, data)}
          prev={prev}
        />
      ),
    },
  ];

  return (
    <>
      <Steps current={currentStep} size="small">
        {steps.map((item) => (
          <Step key={item.title} title={item.title} />
        ))}
      </Steps>
      <br />
      {steps[currentStep].content}
    </>
  );
};

export default DatasetCreateForm;
