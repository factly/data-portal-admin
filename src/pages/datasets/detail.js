import React from 'react';
import { useParams } from 'react-router-dom';
import { useSelector } from 'react-redux';

import DatasetDetails from './components/DatasetDetails';
import DatasetFormatsList from './components/DatasetFormatsList';

const DatasetDetail = () => {
  const { id } = useParams();
  const [datasetSampleId, setDatasetSampleId] = React.useState(null);
  const { dataset, datasetFormats } = useSelector(({ datasets }) => ({
    dataset: datasets.items[id],
    datasetFormats: datasets.items[id]?.formats,
  }));

  React.useEffect(() => {
    if (dataset?.sample_url) {
      const sample = datasetFormats.find((item) => item.url === dataset.sample_url);
      sample && setDatasetSampleId(sample.id);
    }
  }, [datasetFormats]);

  return (
    <>
      <DatasetDetails dataset={dataset} />
      <DatasetFormatsList
        datasetId={dataset.id}
        showOperations={false}
        datasetSampleId={datasetSampleId}
      />
    </>
  );
};

export default DatasetDetail;
