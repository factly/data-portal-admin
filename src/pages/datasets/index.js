import React from 'react';
import DatasetsList from './components/DatasetsList';
import { Button } from 'antd';
import { Link } from 'react-router-dom';

const Datasets = () => {
  return (
    <>
      <Link to={'/datasets/create'}>
        <Button type="primary" style={{ marginBottom: 16 }}>
          Add Dataset
        </Button>
      </Link>
      <DatasetsList />
    </>
  );
};

export default Datasets;
