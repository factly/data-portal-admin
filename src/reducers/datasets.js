import {
  ADD_DATASETS_LIST_REQUEST,
  SET_DATASETS_LIST_CURRENT_PAGE,
  LOADING_DATASETS,
  LOAD_DATASETS_SUCCESS,
  SET_DATASETS_LIST_TOTAL,
  LOAD_DATASETS_FAILURE,
  CREATING_DATASET,
  CREATE_DATASET_SUCCESS,
  CREATE_DATASET_FAILURE,
  UPDATE_DATASET_SUCCESS,
  UPDATE_DATASET_FAILURE,
  DELETE_DATASET_SUCCESS,
  DELETE_DATASET_FAILURE,
} from '../constants/datasets';

const initialState = {
  loading: false,
  ids: [],
  req: [],
  items: {},
  total: 0,
};

export default function datasetsReducer(state = initialState, action = {}) {
  switch (action.type) {
    case LOADING_DATASETS:
      return {
        ...state,
        loading: true,
      };
    case LOAD_DATASETS_SUCCESS: {
      const { items } = action.payload;
      return {
        ...state,
        loading: false,
        items: { ...state.items, ...items },
      };
    }
    case SET_DATASETS_LIST_TOTAL:
      return {
        ...state,
        total: action.payload,
      };
    case ADD_DATASETS_LIST_REQUEST: {
      return {
        ...state,
        req: [...state.req, action.payload],
      };
    }
    case SET_DATASETS_LIST_CURRENT_PAGE:
      return {
        ...state,
        ids: action.payload,
      };
    case LOAD_DATASETS_FAILURE:
      return {
        ...state,
        loading: false,
      };
    case CREATING_DATASET:
      return {
        ...state,
        loading: true,
      };
    case CREATE_DATASET_SUCCESS: {
      const dataset = action.payload;
      return {
        ...state,
        loading: false,
        req: [],
        items: { ...state.items, [dataset.id]: dataset },
        total: state.total + 1,
      };
    }
    case CREATE_DATASET_FAILURE:
      return {
        ...state,
        loading: false,
      };
    case UPDATE_DATASET_SUCCESS: {
      const dataset = action.payload;
      return {
        ...state,
        loading: false,
        items: { ...state.items, [dataset.id]: dataset },
      };
    }
    case UPDATE_DATASET_FAILURE:
      return {
        ...state,
        loading: false,
      };
    case DELETE_DATASET_SUCCESS: {
      const id = action.payload;
      const newItems = { ...state.items };
      delete newItems[id];
      return {
        ...state,
        loading: false,
        req: [],
        ids: [],
        items: newItems,
        total: state.total - 1,
      };
    }
    case DELETE_DATASET_FAILURE:
      return {
        ...state,
        loading: false,
      };
    default:
      return state;
  }
}
