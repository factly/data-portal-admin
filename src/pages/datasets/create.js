import React from 'react';
import { useDispatch } from 'react-redux';
import { useHistory } from 'react-router-dom';

import DatasetCreateForm from './components/DatasetCreateForm';
import { createDataset, createDatasetFormat } from '../../actions/datasets';

const DatasetCreate = () => {
  const history = useHistory();
  const dispatch = useDispatch();

  const onCreate = async (values) => {
    return dispatch(createDataset(values));
    // history.push('/datasets');
  };

  const onCreateDatasetFormat = (datasetId, data) => {
    return dispatch(createDatasetFormat(datasetId, data));
  };

  return (
    <DatasetCreateForm onSubmitDataset={onCreate} onSubmitDatasetFormat={onCreateDatasetFormat} />
  );
};

export default DatasetCreate;
