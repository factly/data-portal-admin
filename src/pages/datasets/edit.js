import React from 'react';
import { useParams, useHistory } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import { Skeleton } from 'antd';

import DatasetCreateForm from './components/DatasetCreateForm';
import { getDataset, updateDataset, createDatasetFormat } from '../../actions/datasets';

const EditDataset = () => {
  const { id } = useParams();
  const history = useHistory();
  const dispatch = useDispatch();
  const { dataset } = useSelector(({ datasets }) => datasets);

  React.useEffect(() => {
    dispatch(getDataset(id));
  }, []);

  if (!dataset.id) return <Skeleton />;

  const onUpdateDataset = (values) => {
    return dispatch(updateDataset(id, values));
    // history.push('/datasets');
  };

  const onCreateDatasetFormat = (datasetId, data) => {
    return dispatch(createDatasetFormat(datasetId, data));
  };

  if (typeof dataset.frequency === 'string') {
    dataset.frequency = {
      count: dataset.frequency?.split(' ')[0],
      units: dataset.frequency?.split(' ')[1],
    };
  }

  return (
    <DatasetCreateForm
      dataset={dataset}
      datasetFormats={dataset.formats}
      onSubmitDataset={onUpdateDataset}
      onSubmitDatasetFormat={onCreateDatasetFormat}
    />
  );
};

export default EditDataset;
