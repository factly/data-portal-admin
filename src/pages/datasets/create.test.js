import React from 'react';
import { useDispatch, Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { mount, shallow } from 'enzyme';
import { act } from '@testing-library/react';

import '../../matchMedia.mock';
import CreateDataset from './create';
import * as actions from '../../actions/datasets';
import DatasetCreateForm from './components/DatasetCreateForm';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn(),
}));

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: jest.fn(),
}));

jest.mock('../../actions/datasets', () => ({
  createDataset: jest.fn(),
  createDatasetFormat: jest.fn(),
}));

describe('Datasets create component', () => {
  let store;
  let mockedDispatch;

  store = mockStore({
    datasets: {
      loading: false,
      ids: [],
      req: [],
      items: {},
      total: 0,
    },
  });
  store.dispatch = jest.fn(() => ({}));
  mockedDispatch = jest.fn(() => Promise.resolve({}));
  useDispatch.mockReturnValue(mockedDispatch);
  describe('snapshot testing', () => {
    it('should render the component', () => {
      const component = shallow(<CreateDataset />);
      expect(component).toMatchSnapshot();
    });
  });
  describe('component testing', () => {
    let wrapper;
    afterEach(() => {
      wrapper.unmount();
    });
    it('should call createDataset', (done) => {
      actions.createDataset.mockReset();
      act(() => {
        wrapper = mount(
          <Provider store={store}>
            <CreateDataset />
          </Provider>,
        );
      });
      wrapper.find(DatasetCreateForm).props().onSubmitDataset({ test: 'test' });
      setTimeout(() => {
        expect(actions.createDataset).toHaveBeenCalledWith({ test: 'test' });
        done();
      }, 0);
    });
    it('should call createDatasetFormat', (done) => {
      actions.createDatasetFormat.mockReset();
      act(() => {
        wrapper = mount(
          <Provider store={store}>
            <CreateDataset />
          </Provider>,
        );
      });
      wrapper.find(DatasetCreateForm).props().onSubmitDatasetFormat(1, { test: 'test' });
      setTimeout(() => {
        expect(actions.createDatasetFormat).toHaveBeenCalledWith(1, { test: 'test' });
        done();
      }, 0);
    });
  });
});
