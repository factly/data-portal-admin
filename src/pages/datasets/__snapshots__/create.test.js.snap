// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Datasets create component snapshot testing should render the component 1`] = `
<DatasetCreateForm
  onSubmitDataset={[Function]}
  onSubmitDatasetFormat={[Function]}
/>
`;
